sdks/java/core/src/main/java/org/apache/beam/sdk/testing/TestPipeline.java

```diff
@@ -292,8 +298,8 @@ public void evaluate() throws Throwable {
   public PipelineResult run() {
     checkState(
         enforcement.isPresent(),
-        "Attempted to run a pipeline while it's enforcement level was not set. Are you "
-            + "using TestPipeline without a @Rule annotation?");
+        "Is your TestPipeline declaration missing a @Rule annotation? Usage: "
```

Comment:
I favor separating those cases as well. It might naturally come from the removal of `PipelineOptions` from pipeline construction - even in testing scenarios, you should be able to full construct a pipeline and later run it using test `PipelineOptions`.

Certainly all the enforcements and test-related functionality of `TestPipeline` are separable from automatically getting `PipelineOptions` from the environment. It has always bothered me that those are coupled, but I understand it is for users to have a single conceptual entry point to "automatically configure and manage this pipeline for testing".